### PR TITLE
download_sysext: integrate signature verification

### PR DESCRIPTION
**Pull request by dongsupark**
Move the delta update part from `test/crau_verify.rs` to `update-format-crau`.
Move `verify_sig` into `update-format-crau`, so it can be imported by delta_update.
Integrate signature verification in `download_sysext`.

### CODE REVIEW

File: src/testdata/omaha-response-example.xml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  <daystart elapsed_seconds="0">
+  </daystart>
+  <app appid="{e96281a6-d1af-4bde-9a0a-97b76e56dc57}" status="ok">
+    <updatecheck status="ok">
+      <urls>
+        <url codebase="https://staging.updateservice.flatcar-linux.net/v1/update/" />
+      </urls>
+      <manifest version="3732.0.0">
+        <packages>
+          <package name="flatcar_production_update.gz" hash_sha256="fa7392bef3a4fa90268b345241d6a3052237ce7d2fa41b4ed0f5e770f3f8dab5" size="382137917" required="false"></package>
```
Comment (pothos) [inline src/testdata/omaha-response-example.xml:12]: Codebase is also wrong, this should be the payload host, not the Omaha endpoint.

How about using this here from the public server for testing:
```
<?xml version="1.0" encoding="UTF-8"?>
<response protocol="3.0" server="nebraska"><daystart elapsed_seconds="0"></daystart><app appid="e96281a6-d1af-4bde-9a0a-97b76e56dc57" status="ok"><ping status="ok"></ping><updatecheck status="ok"><urls><url codebase="https://update.release.flatcar-linux.net/amd64-usr/3732.0.0/"></url></urls><manifest version="3732.0.0"><packages><package name="flatcar_production_update.gz" hash="I6DGNAOrJRUxPbCuLw+AITfiVMo=" size="382137917" required="true"></package><package name="oem-azure.gz" hash="wepxwEV9L2SS1l/ycEZSqWM3dDc=" hash_sha256="3aed3129de50b959a97e4913ba485bd60e72d2bb6aa377d5ed404103f0680043" size="40897503" required="false"></package><package name="oem-qemu.gz" hash="+06iWWI6gaSlcIHV7pjIEJrE9CA=" hash_sha256="8ab630ee4079ecd5f8f512c05b44fec5e4f8db844db916c67c8d54a575cfe506" size="2282" required="false"></package></packages><actions><action event="postinstall" sha256="stLG3U/o4Ar8TMHFwT/RWB0iNkaWOO6QtLrq6+AHBbA=" DisablePayloadBackoff="true"></action></actions></manifest></updatecheck><event status="ok"></event></app></response>
```

Comment (pothos) [reply src/testdata/omaha-response-example.xml:12]: FYI, this is the command that got the above response: `curl -d @<(echo '<?xml version="1.0" encoding="UTF-8"?> <request protocol="3.0" version="update_engine-0.4.10" updaterversion="update_engine-0.4.10" installsource="scheduler" ismachine="1">     <os version="Chateau" platform="CoreOS" sp="2512.2.0_x86_64"></os>     <app appid="e96281a6-d1af-4bde-9a0a-97b76e56dc57" version="1.2.3" track="alpha" bootid="{965fb4c5-ad3e-4eb7-a4c2-ca0c0e31ec84}" oem="ami" oemversion="0.1.1-r1" alephversion="1688.5.3" machineid="abce671d61774703ac7be60715220bfe" lang="en-US" board="amd64-usr" hardware_class="" delta_okay="false" >         <ping active="1"></ping>         <updatecheck></updatecheck>         <event eventtype="3" eventresult="1"></event>     </app> </request>')  -H 'Content-Type: text/xml' https://public.update.flatcar-linux.net/v1/update/`

Comment (dongsupark) [reply src/testdata/omaha-response-example.xml]: Done, thanks

File: src/bin/download_sysext.rs [hunk 2]
```diff
@@ -108,6 +112,34 @@ impl<'a> Package<'a> {
...
+        let res_data = fs::read_to_string(from_path);
+
+        let header = delta_update::read_delta_update_header(&upfile)?;
+
+        // Extract signature from header.
+        let sigbytes = delta_update::get_signatures_bytes(&upfile, &header)?;
+
+        // Parse signature data from the signature containing data, version, special fields.
+        let _sigdata = match delta_update::parse_signature_data(res_data.unwrap().as_bytes(), &sigbytes, PUBKEY_FILE) {
+            Some(data) => Box::leak(data),
```
Comment (wrl) [inline src/bin/download_sysext.rs:131]: don't think we need to `leak()` this

File: src/bin/download_sysext.rs [hunk 3]
```diff
@@ -108,6 +112,34 @@ impl<'a> Package<'a> {
...
+            _ => {
+                self.status = PackageStatus::BadSignature;
+                return Err("unable to parse signature data".into());
+            }
+        };
+
+        println!("Parsed signature data from file {:?}", from_path);
+
+        self.status = PackageStatus::Verified;
```
Comment (wrl) [inline src/bin/download_sysext.rs:140]: do we need a call to `verify_sig_pubkey()` first?

File: src/bin/download_sysext.rs [hunk 4]
```diff
@@ -108,6 +112,34 @@ impl<'a> Package<'a> {
             return true;
         }
     }
+
+    fn verify_signature_on_disk(&mut self, from_path: &Path) -> Result<(), Box<dyn Error>> {
+        const PUBKEY_FILE: &str = "../src/testdata/public_key_test_pkcs8.pem";
```
Comment (wrl) [inline src/bin/download_sysext.rs:117]: should take the path to the pubkey as a function argument rather than having it hard-coded

**Review by wrl — CHANGES_REQUESTED** [review]

File: src/bin/download_sysext.rs (hunk 2 above)
Comment (dongsupark) [reply src/bin/download_sysext.rs]: Oops, good catch. Thanks.

File: src/bin/download_sysext.rs (hunk 4 above)
Comment (dongsupark) [reply src/bin/download_sysext.rs]: Ok, done.

File: src/bin/download_sysext.rs (hunk 3 above)
Comment (dongsupark) [reply src/bin/download_sysext.rs:138]: No, `parse_signature_data()` is already calling `verify_sig_pubkey()`.
https://github.com/flatcar/ue-rs/pull/12/files#diff-4bdabfe18b36b123ce8b3873dae3f39543fc6de6a80b796efddc060d34ef1c60R103

**Comment by dongsupark** [comment]
Improved minor things.
Used argh for parsing input parameters in `test/crau_verify.rs`.

**Pull request merged**